src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -300,7 +300,7 @@ public static Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
 
             if (!s_tokenCache.TryGetValue(targetTime, out tokenTask))
             {
-                var tcs = new TaskCompletionSource<CancellationToken>();
+                var tcs = new TaskCompletionSource<CancellationToken>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
I don't think we can @KKhurin -- that's the known problem with OC.Current and await.  Our current advice to developers is to save to a local and restore after each await.  With the changes in this PR, the awaiters should be guaranteed not to run synchronously with the context of the TCS.TrySetResult() thread, but there are no guarantees they will run on the same thread on which their await began. Asserting OC.Current was the same would trigger unpredictably.  We do need to address our OC.Current vs await story, but I don't believe it should happen in this PR.